pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,437 @@
+	crcontrollerutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	imagev1 "github.com/openshift/api/image/v1"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	controllerutil "github.com/openshift/ci-tools/pkg/controller/util"
+	"github.com/openshift/ci-tools/pkg/util/imagestreamtagmapper"
+	"github.com/openshift/ci-tools/pkg/util/imagestreamtagwrapper"
+)
+
+const ControllerName = "registry_syncer"
+
+func AddToManager(mgr manager.Manager,
+	managers map[string]manager.Manager,
+	pullSecretGetter func() []byte,
+	dryRun bool,
+) error {
+	log := logrus.WithField("controller", ControllerName)
+	r := &reconciler{
+		ctx:              context.Background(),
+		log:              log,
+		registryClients:  map[string]ctrlruntimeclient.Client{},
+		pullSecretGetter: pullSecretGetter,
+		dryRun:           dryRun,
+	}
+	for ctx, m := range managers {
+		r.registryClients[ctx] = imagestreamtagwrapper.MustNew(m.GetClient(), m.GetCache())
+	}
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler: r,
+		// We conflict on ImageStream level which means multiple request for imagestreamtags
+		// of the same imagestream will conflict so stay at one worker in order to reduce the
+		// number of errors we see. If we hit performance issues, we will probably need cluster
+		// and/or imagestream level locking.
+		MaxConcurrentReconciles: 1,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	allContexts := sets.NewString()
+	for ctx := range managers {
+		allContexts.Insert(ctx)
+	}
+
+	for ctx, m := range managers {
+		if err := c.Watch(
+			source.NewKindWithCache(&imagev1.ImageStream{}, m.GetCache()),
+			registryClusterHandlerFactory(sets.NewString(allContexts.List()...).Delete(ctx),
+				testInputImageStreamTagFilterFactory()),
+		); err != nil {
+			return fmt.Errorf("failed to create watch for ImageStreams: %w", err)
+		}
+	}
+
+	r.log.Info("Successfully added reconciler to manager")
+	return nil
+}
+
+type objectFilter func(types.NamespacedName) bool
+
+// registryClusterHandlerFactory produces a handler that:
+// * Watches ImageStreams because ImageStreamTags do not support the watch verb
+// * Extracts all ImageStramTags out of the Image
+// * Filters out the ones that are not in use
+// Note: We can not use a predicate because that is directly applied on the source and the source yields ImageStreams, not ImageStreamTags
+// * Creates a reconcile.Request per cluster and ImageStreamTag
+func registryClusterHandlerFactory(buildClusters sets.String, filter objectFilter) handler.EventHandler {
+	return imagestreamtagmapper.New(func(in reconcile.Request) []reconcile.Request {
+		if !filter(in.NamespacedName) {
+			return nil
+		}
+
+		var requests []reconcile.Request
```

Comment:
We don't need any of this, just `return in` here. One `reconcile.Request` will reconcile both clusters